### PR TITLE
[21362] Fail when trying to serialize `std::string` with null characters on its content

### PR DESCRIPTION
**Pull request by MiguelCompany**
## Description

This addresses #69 and #214 by explicitly failing to serialize a string that contains null characters, instead of silently discard the last part of the string.

@Mergifyio backport 2.1.x 1.0.x
**Note:** I'm not sure whether we should backport this to 1.x.

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- [x] Changes are API compatible. 
- _N/A_ New feature has been added to the `versions.md` file (if applicable).
- [x] Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: CI pass and failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/Cdr.cpp [hunk 1]
```diff
@@ -14,6 +14,7 @@
 
 #include <cstring>
 #include <limits>
+#include <string>
```
Comment (Mario-DL) [inline src/cpp/Cdr.cpp:17]: I think it is no longer needed
```suggestion
```

**Review by Mario-DL — CHANGES_REQUESTED** [review]

**Comment by Mario-DL** [comment]
Should be also make the changes when serializing a `wstring` ?

**Comment by MiguelCompany** [comment]
> Should be also make the changes when serializing a `wstring` ?

Not sure, the standard says that for strings, the terminating null character is not transmitted. Not sure if that implies that the string can have null characters in the middle. I'd say we can handle that on a different PR

**Review by Mario-DL — APPROVED** [review]
LGTM

**Comment by MiguelCompany** [comment]
@Mergifyio backport 2.1.x

**Pull request merged**